jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -1647,4 +1589,149 @@ private void complete()
+                int streamId = frame.getStreamId();
+                if (streamId <= 0)
+                {
+                    synchronized (this)
+                    {
+                        streamId = localStreamIds.getAndAdd(2);
+                        slots.offer(slot);
+                    }
+                    PriorityFrame priority = frame.getPriority();
+                    priority = priority == null ? null : new PriorityFrame(streamId, priority);
+                    frame = new HeadersFrame(streamId, priority, frame);
+                }
+                else
+                {
+                    synchronized (this)
+                    {
+                        slots.offer(slot);
+                    }
+                }
+
+                fillSlotWithNewLocalStream(slot, frame, promise, listener, streamId);
+            }
+            catch (Throwable x)
+            {
+                removeSlotAndFailPromise(slot, promise, x);
+            }
+        }
+
+        private void push(PushPromiseFrame frame, Promise<Stream> promise, Stream.Listener listener)
+        {
+            Slot slot = new Slot();
+            try
+            {
+                int streamId;
+                synchronized (this)
+                {
+                    streamId = localStreamIds.getAndAdd(2);
+                    slots.offer(slot);
+                }
+                frame = new PushPromiseFrame(streamId, frame);
+
+                fillSlotWithNewLocalStream(slot, frame, promise, listener, streamId);
+            }
+            catch (Throwable x)
+            {
+                removeSlotAndFailPromise(slot, promise, x);
+            }
+        }
+
+        private void fillSlotWithNewLocalStream(Slot slot, Frame frame, Promise<Stream> promise, Stream.Listener listener, int streamId)
+        {
+            IStream stream = createLocalStream(streamId);
+            stream.setListener(listener);
+            slot.entry = new ControlEntry(frame, stream, new StreamPromiseCallback(promise, stream));
+            flush();
+        }
+
+        private void removeSlotAndFailPromise(Slot slot, Promise<Stream> promise, Throwable x)
+        {
+            synchronized (this)
+            {
+                slots.remove(slot);
+            }
+            flush();
+            promise.failed(x);
+        }
+
+        private void flush()
+        {
+            Thread thread = Thread.currentThread();
+            boolean queued = false;
+            while (true)
+            {
+                ControlEntry entry = null;
+                synchronized (this)
+                {
+                    if (flushing != null && flushing != thread)
+                        return;
+                    flushing = thread;
+
+                    Slot slot = slots.peek();
+                    if (slot != null)
+                        entry = slot.entry;
+                    if (entry == null)
+                    {
+                        flushing = null;
```

Comment:
So if another thread is just about the set the entry, then it will immediately call `flush` afterwards. Because `flush` grabs the lock, it will either be blocked while this thread exits from here, or this thread will have completed.... either way, it will be the other thread that continues flushing.
Maybe some comments to well explain that.

Reply:
you're right, I added a comment describing that mechanism.